interactions/api/gateway/ratelimit.py

```diff
@@ -0,0 +1,77 @@
+import asyncio
+import logging
+from sys import version_info
+from time import time
+from typing import Optional
+
+log = logging.getLogger("gateway.ratelimit")
+
+
+class WSRateLimit:
+    """
+    A class that controls Gateway ratelimits using locking and a timer.
+
+    .. note ::
+        While the docs state that the Gateway ratelimits are 120/60 (120 requests per 60 seconds),
+        this ratelimit offsets to 115 instead of 120 for room.
+
+    :ivar Lock lock: The gateway Lock object.
+    :ivar int max: The upper limit of the ratelimit. Defaults to `115` seconds.
+    :ivar int remaining: How many requests are left per ``per_second``. This is automatically decremented and reset.
+    :ivar float current_limit: When this cooldown session began. This is defined automatically.
+    :ivar float per_second: A constant denoting how many requests can be done per unit of seconds. (i.e., per 60 seconds, per 45, etc.)
+    """
+
+    def __init__(self, loop: Optional[asyncio.AbstractEventLoop] = None):
+        self.lock = asyncio.Lock(loop=loop) if version_info < (3, 10) else asyncio.Lock()
+        # To conserve timings, we need to do 115/60
+        # Also, credit to d.py for their ratelimiter inspiration
+
+        self.max = self.remaining = 115
+        self.per_second = 60.0
+        self.current_limit = 0.0
+
+    def is_ratelimited(self) -> bool:
```

Comment:
Why don't we make this and `get_delay` a property and call it `ratelimited` and `delay`? It doesn't have to be it just looks better imo